rai/node/bootstrap.cpp

```diff
@@ -950,12 +960,78 @@ bool rai::bootstrap_attempt::consume_future (std::future<bool> & future_a)
 
 void rai::bootstrap_attempt::process_fork (MDB_txn * transaction_a, std::shared_ptr<rai::block> block_a)
 {
+	std::unique_lock<std::mutex> lock (mutex);
+	try_resolve_fork (transaction_a, block_a, true);
+}
+
+void rai::bootstrap_attempt::try_resolve_fork (MDB_txn * transaction_a, std::shared_ptr<rai::block> block_a, bool from_processor)
+{
+	assert (!mutex.try_lock ());
+	std::weak_ptr<rai::bootstrap_attempt> this_w (shared_from_this ());
 	std::shared_ptr<rai::block> ledger_block (node->ledger.forked_block (transaction_a, *block_a));
-	if (!node->active.start (transaction_a, ledger_block))
+	if (ledger_block)
 	{
-		node->network.broadcast_confirm_req (ledger_block);
-		node->network.broadcast_confirm_req (block_a);
-		BOOST_LOG (node->log) << boost::str (boost::format ("While bootstrappping, fork between our block %1% and block %2% both with root %3%") % ledger_block->hash ().to_string () % block_a->hash ().to_string () % block_a->root ().to_string ());
+		node->active.start (transaction_a, ledger_block, [this_w, block_a](std::shared_ptr<rai::block>, bool resolved) {
+			if (auto this_l = this_w.lock ())
+			{
+				if (resolved)
+				{
+					{
+						std::unique_lock<std::mutex> lock (this_l->mutex);
+						this_l->unresolved_forks.erase (block_a->hash ());
+						this_l->condition.notify_all ();
+					}
+					rai::transaction transaction (this_l->node->store.environment, nullptr, false);
```

Comment:
IIUC this is almost certain to trigger I/O.

Reply:
Ok, Luke showed me I was missing the fact that this block is all deferred work in a lambda.

Reply:
I reduced the scope of the mutex anyways, it wasn't ideal.